File: frontend/src/api/fleet-insights.ts
import { getFleetComparisonApiV1ReportsInsightsFleetComparisonGet } from "../generated/api/reports-v2/reports-v2";

export type PeriodType = "week" | "month";

export interface PeriodMetrics {
  fuel_l: number;
  fuel_cost_tl: number;
  anomaly_count: number;
  trip_count: number;
}

export interface FleetComparisonResponse {
  period: PeriodType;
  current: PeriodMetrics;
  previous: PeriodMetrics;
  fuel_l_delta_pct: number | null;
  fuel_cost_delta_pct: number | null;
  anomaly_delta_pct: number | null;
  trip_delta_pct: number | null;
  current_start: string;
  current_end: string;
  previous_start: string;
  previous_end: string;
}

export const fleetInsightsService = {
  getComparison: async (
    period: PeriodType = "month",
  ): Promise<FleetComparisonResponse> => {
    const result =
      await getFleetComparisonApiV1ReportsInsightsFleetComparisonGet({
        period,
      });
    return result as unknown as FleetComparisonResponse;
  },
};

File: app/schemas/fleet_insights.py
"""Reports v2 RV2.2 — Fleet İçgörü schemas."""

from __future__ import annotations

from datetime import date
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

PeriodType = Literal["week", "month"]


class PeriodMetricsResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    fuel_l: float = Field(..., ge=0)
    fuel_cost_tl: float = Field(..., ge=0)
    anomaly_count: int = Field(..., ge=0)
    trip_count: int = Field(..., ge=0)


class FleetComparisonResponse(BaseModel):
    """Period-over-period karşılaştırma çıktısı."""

    model_config = ConfigDict(from_attributes=True)

    period: PeriodType
    current: PeriodMetricsResponse
    previous: PeriodMetricsResponse
    fuel_l_delta_pct: Optional[float] = Field(
        None, description="None ise önceki periyot 0 idi"
    )
    fuel_cost_delta_pct: Optional[float] = None
    anomaly_delta_pct: Optional[float] = None
    trip_delta_pct: Optional[float] = None
    current_start: date
    current_end: date
    previous_start: date
    previous_end: date
